cmake/HPX_AddModule.cmake

```diff
@@ -0,0 +1,181 @@
+
+  hpx_debug("Add module ${name}: SOURCE_ROOT: ${SOURCE_ROOT}")
+  hpx_debug("Add module ${name}: HEADER_ROOT: ${HEADER_ROOT}")
+
+  if(HPX_${name_upper}_WITH_COMPATIBILITY_HEADERS)
+    set(COMPAT_HEADER_ROOT "${CMAKE_CURRENT_SOURCE_DIR}/include_compatibility")
+    hpx_debug("Add module ${name}: COMPAT_HEADER_ROOT: ${COMPAT_HEADER_ROOT}")
+  endif()
+
+  # Write full path for the sources files
+  include(HPX_CMakeUtils)
+  prepend(sources ${SOURCE_ROOT} ${${name}_SOURCES})
+  prepend(headers ${HEADER_ROOT} ${${name}_HEADERS})
+  if(HPX_${name_upper}_WITH_COMPATIBILITY_HEADERS)
+    prepend(compat_headers ${COMPAT_HEADER_ROOT} ${${name}_COMPAT_HEADERS})
+  endif()
+
+  # This header generation is disabled for config module specific generated
+  # headers are included
+  if (${name}_GLOBAL_HEADER_GEN)
+    # Add a global include file that include all module headers
+    FILE(WRITE ${CMAKE_BINARY_DIR}/hpx/${name}.hpp
+      "//  Copyright (c) 2019 The STE||AR GROUP\n"
+      "//\n"
+      "//  Distributed under the Boost Software License, Version 1.0. (See accompanying\n"
+      "//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)\n\n"
+      "#ifndef HPX_${name_upper}_HPP\n"
+      "#define HPX_${name_upper}_HPP\n\n"
+    )
+    foreach(header_file ${${name}_HEADERS})
+      FILE(APPEND ${CMAKE_BINARY_DIR}/hpx/${name}.hpp
+        "#include <${header_file}>\n"
+      )
+    endforeach(header_file)
+    FILE(APPEND ${CMAKE_BINARY_DIR}/hpx/${name}.hpp
+      "\n#endif\n"
+    )
+  endif()
+
+  foreach(header_file ${headers})
+    hpx_debug(${header_file})
+  endforeach(header_file)
+
+  add_library(hpx_${name} STATIC ${sources} ${headers} ${compat_headers})
+
+  target_link_libraries(hpx_${name} ${${name}_DEPENDENCIES})
+  target_include_directories(hpx_${name} PUBLIC
+    $<BUILD_INTERFACE:${HEADER_ROOT}>
+    $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}>)
+
+  if(HPX_${name_upper}_WITH_COMPATIBILITY_HEADERS)
+    target_include_directories(hpx_${name} PUBLIC
+      $<BUILD_INTERFACE:${COMPAT_HEADER_ROOT}>)
+  endif()
+
+  target_compile_definitions(hpx_${name} PRIVATE
+    $<$<CONFIG:Debug>:DEBUG>
+    $<$<CONFIG:Debug>:_DEBUG>
+    HPX_MODULE_EXPORTS
+  )
+
+  add_hpx_source_group(
+    NAME hpx_{name}
+    ROOT ${HEADER_ROOT}/hpx
+    CLASS "Header Files"
+    TARGETS ${headers})
+  add_hpx_source_group(
+    NAME hpx_{name}
+    ROOT ${SOURCE_ROOT}
+    CLASS "Source Files"
+    TARGETS ${sources})
+  if(HPX_${name_upper}_WITH_COMPATIBILITY_HEADERS)
+    add_hpx_source_group(
+      NAME hpx_${name}
+      ROOT ${COMPAT_HEADER_ROOT}/hpx
+      CLASS "Header Files"
+      TARGETS ${compat_headers})
+  endif()
+
+  set_target_properties(hpx_${name} PROPERTIES
+    FOLDER "Core/Modules")
+
+  install(TARGETS hpx_${name} EXPORT HPXTargets
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    COMPONENT ${name}
+  )
```

Comment:
Should we really install the generated binaries? I thought we need those while building the HPX core library only/

Reply:
Let's see with @msimberg when he returns, but installing the module libraries could simplify the compilation process if we consider the modules as standalone in the future, just linking them from the binaries, but as I said, I'm not sure of what mickael wants to do with it ^^

Reply:
I'm not 100% sure where I want to go with this but having them installed for now won't hurt. Basically we can keep it or remove it now and it won't make a difference because users shouldn't be linking manually to these anyway. Eventually I would like at least a few high-level shared libraries that can be used explicitly. Exposing all the small static libraries would be nice as well, but we'd need to make sure that things work correctly when linking with a static library and a shared library that includes that static library (probably won't work).

Reply:
What if we made this optional using a `INSTALL_BINARIES` tag for the `add_hpx_module()` function (default could be `OFF` for now)? I'm a bit hesitant to install those now as users might think using them this is a supported scenario... - I know this is more wishful thinking than anything else - but anyways.